driver-core/src/main/java/com/datastax/driver/core/AbstractAddressableByIndexData.java

```diff
@@ -43,75 +48,111 @@ protected ByteBuffer getValue(int i) {
     }
 
     public T setTimestamp(int i, Date v) {
-        checkType(i, DataType.Name.TIMESTAMP);
-        return setValue(i, v == null ? null : TypeCodec.TimestampCodec.instance.serialize(v));
+        checkType(i, TIMESTAMP);
+        return setValue(i, v == null ? null : codecFor(i, Date.class).serialize(v, protocolVersion));
     }
 
     public T setDate(int i, LocalDate v) {
-        checkType(i, DataType.Name.DATE);
-        return setValue(i, TypeCodec.DateCodec.instance.serialize(v));
+        checkType(i, DATE);
+        return setValue(i, v == null ? null : codecFor(i, LocalDate.class).serialize(v, protocolVersion));
     }
 
     public T setTime(int i, long v) {
-        checkType(i, DataType.Name.TIME);
-        return setValue(i, TypeCodec.TimeCodec.instance.serializeNoBoxing(v));
+        checkType(i, TIME);
+        TypeCodec<Long> codec = codecFor(i, Long.class);
+        ByteBuffer bb;
+        if(codec instanceof TypeCodec.PrimitiveLongCodec)
+            bb = ((TypeCodec.PrimitiveLongCodec)codec).serializeNoBoxing(v, protocolVersion);
+        else
+            bb = codec.serialize(v, protocolVersion);
+        return setValue(i, bb);
     }
 
     public T setFloat(int i, float v) {
-        checkType(i, DataType.Name.FLOAT);
-        return setValue(i, TypeCodec.FloatCodec.instance.serializeNoBoxing(v));
+        checkType(i, FLOAT);
+        TypeCodec<Float> codec = codecFor(i, Float.class);
+        ByteBuffer bb;
+        if(codec instanceof TypeCodec.PrimitiveFloatCodec)
+            bb = ((TypeCodec.PrimitiveFloatCodec)codec).serializeNoBoxing(v, protocolVersion);
+        else
+            bb = codec.serialize(v, protocolVersion);
+        return setValue(i, bb);
     }
 
     public T setDouble(int i, double v) {
-        checkType(i, DataType.Name.DOUBLE);
-        return setValue(i, TypeCodec.DoubleCodec.instance.serializeNoBoxing(v));
+        checkType(i, DOUBLE);
+        TypeCodec<Double> codec = codecFor(i, Double.class);
+        ByteBuffer bb;
+        if(codec instanceof TypeCodec.PrimitiveDoubleCodec)
+            bb = ((TypeCodec.PrimitiveDoubleCodec)codec).serializeNoBoxing(v, protocolVersion);
+        else
+            bb = codec.serialize(v, protocolVersion);
+        return setValue(i, bb);
     }
 
     public T setString(int i, String v) {
-        DataType.Name type = checkType(i, DataType.Name.VARCHAR, DataType.Name.TEXT, DataType.Name.ASCII);
-        switch (type) {
-            case ASCII:
-                return setValue(i, v == null ? null : TypeCodec.StringCodec.asciiInstance.serialize(v));
-            case TEXT:
-            case VARCHAR:
-                return setValue(i, v == null ? null : TypeCodec.StringCodec.utf8Instance.serialize(v));
-            default:
-                throw new AssertionError();
-        }
+        checkType(i, VARCHAR, TEXT, ASCII);
+        return setValue(i, v == null ? null : codecFor(i, String.class).serialize(v, protocolVersion));
```

Comment:
Won't this be a problem in that we won't be able to distinguish which codec to use between ascii/utf8 for serialization/deserlization?  Seems like now it would always use ascii codec by default based on the CodecRegistry.